gwcs/wcs.py

```diff
@@ -634,3 +641,211 @@ def fix_inputs(self, fixed):
+    # The case of one pass with the specified polynomial degree
+    if degree is None:
+        deglist = range(1,10)
+    else:
+        deglist = [degree]
+    for deg in deglist:
+        poly_x = Polynomial2D(degree=deg)
+        poly_y = Polynomial2D(degree=deg)
+        fit_poly_x = llsqfitter(poly_x, u, v, x)
+        fit_poly_y = llsqfitter(poly_y, u, v, y)
+        max_resid = _compute_distance_residual(x, y,
+                                               fit_poly_x(u, v), fit_poly_y(u, v))
+        if verbose:
+            print(f'Degree = {degree}, max_resid = {max_resid}')
+        if max_error is not None and max_resid < max_error:
+            if verbose:
+                print('terminating condition met')
+            break
+        if degree is None and deg == 9:
+            raise RuntimeError("Unable to achieve required fit accuracy with SIP polynomials")
+    return fit_poly_x, fit_poly_y, max_resid
+
+
+def _compute_distance_residual(undist_x, undist_y, fit_poly_x, fit_poly_y):
+    """
+    Compute the distance residuals and return the rms and maximum values.
+    """
+    dist = np.sqrt((undist_x - fit_poly_x)**2 + (undist_y - fit_poly_y)**2)
+    max_resid = dist.max()
+    return max_resid
+
+
+def _reform_poly_coefficients(fit_poly_x, fit_poly_y):
+    """
+    The fit polynomials must be recombined to align with the SIP decomposition
+
+    The result is the f(u,v) and g(u,v) polynomials, and the CD matrix.
+    """
+    # Extract values for CD matrix and recombining
+    c11 = fit_poly_x.c1_0.value
+    c12 = fit_poly_x.c0_1.value
+    c21 = fit_poly_y.c1_0.value
+    c22 = fit_poly_y.c0_1.value
+    sip_poly_x = fit_poly_x.copy()
+    sip_poly_y = fit_poly_y.copy()
+    # Force low order coefficients to be 0 as defined in SIP
+    sip_poly_x.c0_0 = 0
+    sip_poly_y.c0_0 = 0
+    sip_poly_x.c1_0 = 0
+    sip_poly_x.c0_1 = 0
+    sip_poly_y.c1_0 = 0
+    sip_poly_y.c0_1 = 0
+
+    cdmat = ((c11, c12), (c21, c22))
+    invcdmat = npla.inv(np.array(cdmat))
+    degree = fit_poly_x.degree
+    # Now loop through all remaining coefficients
+    for i in range(0, degree + 1):
+        for j in range(0, degree + 1):
+            if (i + j > 1) and (i + j < degree + 1):
+                old_x = getattr(fit_poly_x, f'c{i}_{j}').value
+                old_y = getattr(fit_poly_y, f'c{i}_{j}').value
+                newcoeff = np.dot(invcdmat, np.array([[old_x], [old_y]]))
+                setattr(sip_poly_x, f'c{i}_{j}', newcoeff[0, 0])
+                setattr(sip_poly_y, f'c{i}_{j}', newcoeff[1, 0])
+
+    return cdmat, sip_poly_x, sip_poly_y
+
+
+def _store_2D_coefficients(hdr, poly_model, coeff_prefix, keeplinear=False):
+    """
+    Write the polynomial model coefficients to the header.
+    """
+    mindeg = int(not keeplinear)
+    degree = poly_model.degree
+    for i in range(0, degree + 1):
+        for j in range(0, degree + 1):
+            if (i + j) > mindeg and (i + j < degree + 1):
```

Comment:
Same thing here. I feel this kind of loop and `if` could be simplified with something like:
```suggestion
    for i in range(degree + 1):
        for j in range(degree + 1 - i):
            if i + j > mindeg:
```